src/policy/discount.h

```diff
@@ -24,15 +24,21 @@ static inline int64_t GetDiscountVirtualTransactionSize(const CTransaction& tx,
     // for each confidential output
     for (size_t i = 0; i < tx.vout.size(); ++i) {
         const CTxOut& output = tx.vout[i];
-        if (output.IsConfidential()) {
-            // subtract the weight of the output witness
-            if (i < tx.witness.vtxoutwit.size()) {
-                weight -= ::GetSerializeSize(tx.witness.vtxoutwit[i], PROTOCOL_VERSION);
+        if (i < tx.witness.vtxoutwit.size()) {
+            if (output.nValue.IsCommitment()) {
+                // subtract the weight of the output value range proof
+                weight -= ::GetSerializeSize(tx.witness.vtxoutwit[i].vchRangeproof, PROTOCOL_VERSION);
```

Comment:
Move the `if (i < tx.witness.vtxoutwit.size())` inside and only apply it to the `::GetSerializeSize(tx.witness.vtxoutwit[i].vchRangeproof, PROTOCOL_VERSION);` statement.

Reply:
Scratch that.

You should go back and discount the entire `::GetSerializeSize(tx.witness.vtxoutwit[i], PROTOCOL_VERSION);` whenever `i < tx.witness.vtxoutwit.size()` independent of whether or not any assets or values are blinded.

At least that is what I think.  @apoelstra thoughts?

Reply:
Yeah, it also occurred to me that it would be simpler to separate things out this way. So then each individual discount is applied exactly if it makes sense.

Though I *think* that in practice the CT rules mean that the existing logic is correct.

Reply:
One issue with the current code is that if you have a transaction with only a confidential nonce, but it has no witness data, then that confidential nonce doesn't get a discount.  ... not that such transactions exist in practice, but still.

Reply:
sorry - I did mean to have the nonce check outside of the witness check since it's unrelated. pushed a fixup commit for that. 

the rest of this thread is not clear to me. as it stands now, the range proof for value is discounted if the amount is blinded, and the surjection proof for asset is discounted if the asset is blinded. 

please specify any change I should make now?

Reply:
>the rest of this thread is not clear to me. as it stands now, the range proof for value is discounted if the amount is blinded, and the surjection proof for asset is discounted if the asset is blinded.

This is fine, but you don't need to make it conditional. Just always discount the surjectionproof and rangeproof (as long as the vtxoutwit is big enough). Yes, it is true that these fields must be empty unless the value or asset respectively are confidential. But we don't need to rely on that fact here, so let's avoid relying on it.

Reply:
Interestingly, discounting the witness unconditionally affects explicit transactions by reporting their `discountvsize` as 2 bytes smaller than their `vsize` (324 vs 326 in the test case tx). Presumably that's a serialized `0` for each of the two proofs.

Reply:
Ah, yes. These fields have a serialized length byte even if they're unpopulated.

This means your code should have a `- 1` because you *shouldn't* be discounting it because it's present even in non-CT transactions. We found a bug :).